File: Tree-shaped Deep Forest/Binary Tree/Q2. Create Binary Tree From Descriptions/Q2_CreateBinaryTreeFromDescriptions.java
class Solution {
    public TreeNode createBinaryTree(int[][] descriptions) {

        Set<Integer> childrenSet = new HashSet<>(); // store all nodes that appear as children
        Map<Integer, int[]> childrenHashmap = new HashMap<>(); // parent -> [leftChild, rightChild]

        for (int[] desc : descriptions) {
            int parent = desc[0];          // parent node value
            int child = desc[1];           // child node value
            boolean isLeft = desc[2] == 1; // 1 means left child, 0 means right child

            // create entry for parent if not present
            childrenHashmap.putIfAbsent(parent, new int[]{-1, -1});

            childrenSet.add(child); // record child nodes to help find root later

            if (isLeft) {
                childrenHashmap.get(parent)[0] = child; // set left child
            } else {
                childrenHashmap.get(parent)[1] = child; // set right child
            }
        }

        int headNodeVal = 0; // will store root node value

        // find root (node that never appears as a child)
        for (int parent : childrenHashmap.keySet()) {
            if (!childrenSet.contains(parent)) {
                headNodeVal = parent;
                break;
            }
        }

        // build tree starting from root
        return constructTree(headNodeVal, childrenHashmap);
    }

    private TreeNode constructTree(int curNodeVal, Map<Integer, int[]> childrenHashmap) {

        TreeNode newNode = new TreeNode(curNodeVal); // create node

        // check if this node has children
        if (childrenHashmap.containsKey(curNodeVal)) {

            int[] children = childrenHashmap.get(curNodeVal); // get left and right child values

            if (children[0] != -1) {
                // recursively build left subtree
                newNode.left = constructTree(children[0], childrenHashmap);
            }

            if (children[1] != -1) {
                // recursively build right subtree
                newNode.right = constructTree(children[1], childrenHashmap);
            }
        }

        return newNode; // return constructed subtree
    }
}

File: Tree-shaped Deep Forest/Binary Tree/Q2. Create Binary Tree From Descriptions/Q2_CreateBinaryTreeFromDescriptions.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val      # value of the node
#         self.left = left    # reference to left child
#         self.right = right  # reference to right child

class Solution(object):
    def createBinaryTree(self, descriptions):
        """
        :type descriptions: List[List[int]]
        :rtype: Optional[TreeNode]
        """

        # recursive function to build tree from a node value
        def construct_tree(cur_node_val):

            new_node = TreeNode(cur_node_val)  # create node

            # check if this node has children
            if cur_node_val in children_hashmap:

                # if left child exists, recursively build left subtree
                if children_hashmap[cur_node_val][0] is not None:
                    new_node.left = construct_tree(children_hashmap[cur_node_val][0])

                # if right child exists, recursively build right subtree
                if children_hashmap[cur_node_val][1] is not None:
                    new_node.right = construct_tree(children_hashmap[cur_node_val][1])

            return new_node  # return constructed subtree


        children_set = set()      # store all nodes that appear as children
        children_hashmap = {}     # parent -> [leftChild, rightChild]

        # process each description
        for parent, child, isleft in descriptions:

            # create entry for parent if not present
            if parent not in children_hashmap:
                children_hashmap[parent] = [None, None]  # [left, right]

            children_set.add(child)  # record child node

            # decide whether child is left or right
            target = 0 if isleft else 1

            # assign child to left or right position
            children_hashmap[parent][target] = child


        # find root (node that never appears as a child)
        for parent in children_hashmap:
            if parent not in children_set:
                head_node_val = parent
                break

        # construct tree starting from root
        head = construct_tree(head_node_val)

        return head  # return root of constructed tree